### PR TITLE
[WIP] Allow mods to provide additional package loaders.

### PR DESCRIPTION
**Pull request by IceReaper**
See TODO in OpenRA.Game/ModData.cs - im not quite shure whats the best way to get around this.
If anyone has ideas on how to improve or finish this PR, i would like to read your feedback.

### CODE REVIEW

File: OpenRA.Game/FileSystem/ZipFile.cs [hunk 1]
```diff
@@ -19,6 +19,12 @@
 
 namespace OpenRA.FileSystem
 {
+	public class ZipFileExtensions : IExtensionBasedPackage
+	{
+		public IEnumerable<string> Extensions { get { return new []{"zip", "oramap", "oramod"}; } }
+		public IReadOnlyPackage Create(FileSystem context, string filename) { return new MixFile(context, filename); }
```
Comment (GraionDilach) [inline OpenRA.Game/FileSystem/ZipFile.cs]: new MixFile?!

File: OpenRA.Game/FileSystem/ZipFile.cs [hunk 2]
```diff
@@ -19,6 +19,12 @@
 
 namespace OpenRA.FileSystem
 {
+	public class ZipFileExtensions : IExtensionBasedPackage
+	{
+		public IEnumerable<string> Extensions { get { return new []{"zip", "oramap", "oramod"}; } }
```
Comment (GraionDilach) [inline OpenRA.Game/FileSystem/ZipFile.cs]: Oramod is gone.

**Comment by GraionDilach** [comment]
Couldn't you just introduce a PackageFormats in mod.yaml (similarly to SoundFormats) where the mod is required to list all intended package extensions it plans to use and then resolve those types instead? That sounds like a much better strategy to me albeit I haven't looked at yet why it isn't tthe current way.

**Review by pchote — CHANGES_REQUESTED** [review]
Can you please make this follow the same pattern as `ISpriteLoader` (e.g. [ShpTSLoader](https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/SpriteLoaders/ShpTSLoader.cs)) and `ISoundLoader` (e.g. [WavLoader](https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/AudioLoaders/WavLoader.cs))?  You can use `GetLoaders<IPackageLoader>` in [ModData](https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Game/ModData.cs#L71) to initialize them.

For simplicity I would make the `bool IPackageLoader.TryParsePackage(Stream stream, string name, out IReadOnlyPackage package)` accept a filename as well as the stream so that we can, for now, check against extension.  In the future they can then migrate to doing proper data type checks.

File: OpenRA.Game/FileSystem/ZipFile.cs (hunk 1 above)
Comment (IceReaper) [inline OpenRA.Game/FileSystem/ZipFile.cs]: Ah, thanks for that, its a WIP, copy and pasted that. already fixed locally.

File: OpenRA.Game/FileSystem/ZipFile.cs (hunk 2 above)
Comment (IceReaper) [inline OpenRA.Game/FileSystem/ZipFile.cs]: Was still present in original code, thats why i kept that.

**Comment by IceReaper** [comment]
This implementation works.
However i got some questions left:

1. which mods do actualy require which PackageLoader to work? The D2kSoundFile is self-explaining...
2. there is an InstallShieldFile class. I guess its only used by the modloader mod, as it extracts those?
3. any more thoughts or feedback on this?

**Comment by IceReaper** [comment]
Open topics:
- cnc is the only mod which seems to not load due to snow.mix problems.
- A lot of empty folders are created if a package cannot be found.
- Doing ```ModFiles.LoadFromManifest(Manifest);``` twice is ugly and need to be changed.

**Comment by pchote** [comment]
I have some ideas on how to solve those in a way that also fixes @abcdefg30's SharpDevelop issues without breaking our engine design.  I plan to adapt the code in this PR and push an improved version, but it may take a few days until I can get around to it.

**Comment by pchote** [comment]
I haven't forgotten about this.  I just haven't found a solution that I am happy with yet.

**Comment by pchote** [comment]
I've expanded the scope of this a bit, and filed a new pr #13223.  Thanks for giving me a headstart on this!

**Pull request closed without merging**